### PR TITLE
Add `NoInfer` intrinsic type

### PR DESCRIPTION
**Pull request by Andarist**
closes https://github.com/microsoft/TypeScript/issues/14829

There is still some work to be done here (type displays on hover, assignability bug, and maybe more) - but I'm pushing this out to receive an early feedback

cc @RyanCavanaugh

### CODE REVIEW

File: tests/baselines/reference/noInfer.errors.txt [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    
+    declare class ExampleClass<T> {}
+    class OkClass<T> {
+        constructor(private clazz: ExampleClass<T>, private _value: NoInfer<T>) {}
+    
+        get value(): T {
+            return this._value; // ok
+            ~~~~~~~~~~~~~~~~~~~
+!!! error TS2322: Type 'NoInfer<T>' is not assignable to type 'T'.
+!!! error TS2322:   'T' could be instantiated with an arbitrary type which could be unrelated to 'NoInfer<T>'.
```
Comment (Andarist) [inline tests/baselines/reference/noInfer.errors.txt:80]: This should be OK - gonna fix this soon

**Comment by sandersn** [comment]
@Andarist did you learn anything interesting from the implementation? If so, can you add a report to the end of #14829? Unless @DanielRosenwasser @RyanCavanaugh or @ahejlsberg are interested in taking up this feature in a design meeting, it's likely that this PR will sit until it gets stale.

**Comment by Andarist** [comment]
@jakebailey would you mind creating a playground for this? :)

**Comment by jakebailey** [comment]
@typescript-bot pack this

**Comment by unional** [comment]
Hi, I am reading https://github.com/microsoft/TypeScript/issues/14829 and trying to add `NoInfer<T>` to fix the `stub()` function in [type-plus](https://github.com/unional/type-plus).

An I notice the type `type NoInfer<T> = T & { [K in keyof T]: T[K] }` there works with the `Animal|Dog` example but does not work with:

```ts
type NoInfer<T> = T & { [K in keyof T]: T[K] }

type RecursivePartial<T> = {
	[P in keyof T]?: T[P] extends (infer U)[]
		? RecursivePartial<U>[]
		: T[P] extends Record<any, any>
		? RecursivePartial<T[P]>
		: T[P]
}

function foo(v: { a: number, b: string }) {}
function stub<T>(v: RecursivePartial<NoInfer<T>>): T { return v as T }

foo(stub({ a: 1 }) // error

// Argument of type '{ a: number; }' is not assignable to parameter of type '{ a: number; b: string; }'.
// Property 'b' is missing in type '{ a: number; }' but required in type '{ a: number; b: string; }'.
```

It works with the following type: `type NoInfer<T> = Or<IsNull<T>, IsUndefined<T>, T, T & {}>`
(where `Or<A,B,Then,Else>`, `IsNull<T>`, `IsUndefined<T>` does what their name suggest).

Just want to bring it up so that this test case is covered in your implementation.
🌷

**Comment by andrewbranch** [comment]
@typescript-bot pack this

**Comment by Andarist** [comment]
superseded by https://github.com/microsoft/TypeScript/pull/56794

**Pull request closed without merging**